### PR TITLE
Devirtualize -If and vectorize count

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Now the `-If` combinator is devirtualized, and `count` is properly vectorized. This is for https://github.com/ClickHouse/ClickHouse/pull/17041 

Detailed description / Documentation draft:

There are other cases to be optimized. We can also fulfill https://github.com/ClickHouse/ClickHouse/issues/16330

### CODE REVIEW

**Comment by amosbird** [comment]
Test case :  `select countIf(1) from numbers(1000000000) settings max_threads = 1`

Baseline: 3.1 sec
Devirtualize `-If`:  2.3 sec
Devirtualize and vectorize: 0.4 sec

**Comment by alexey-milovidov** [comment]
Translation units became too heavy, they cannot build under the limits:
```
prlimit --rss=10000000 --cpu=600
```

Let's split them to multiple translation units.

File: src/AggregateFunctions/IAggregateFunction.h [hunk 1]
```diff
@@ -217,31 +236,90 @@ class IAggregateFunctionHelper : public IAggregateFunction
...
+        size_t batch_size,
+        AggregateDataPtr * places,
+        size_t place_offset,
+        const IColumn ** columns,
+        Arena * arena,
+        size_t num_arguments = 0) const override
     {
-        for (size_t i = 0; i < batch_size; ++i)
-            static_cast<const Derived *>(this)->add(places[i] + place_offset, columns, i, arena);
+        if (num_arguments > 0)
```
Comment (alexey-milovidov) [inline src/AggregateFunctions/IAggregateFunction.h:247]: This looks strange. What if the function already has multiple arguments, but there is no `-If` combinator?

Example: `corr`.

Comment (amosbird) [reply src/AggregateFunctions/IAggregateFunction.h:247]: > This looks strange. What if the function already has multiple arguments, but there is no -If combinator?

It's indeed bad name.  `num_arguments` is served as an `-If` indicator here, and it denotes the pos of `-If` cond, which equals to the nested_func's num_arguments

Comment (amosbird) [reply src/AggregateFunctions/IAggregateFunction.h:247]: And the only special no-argument agg: `count` has specialized implementation.

Comment (alexey-milovidov) [reply src/AggregateFunctions/IAggregateFunction.h:247]: Let's name it `mask_argument_pos`, it can be ssize_t with -1 as no value.

**Comment by alexey-milovidov** [comment]
> AST fuzzer — Assertion `n < size()' failed 

Error with Map datatype - already fixed in master.

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**